flink-runtime/src/main/java/org/apache/flink/runtime/metrics/DescriptiveStatisticsHistogramStatistics.java

```diff
@@ -19,53 +19,174 @@
  */
 public class DescriptiveStatisticsHistogramStatistics extends HistogramStatistics {
-	private final DescriptiveStatistics descriptiveStatistics;
+	private final CommonMetricsSnapshot statisticsSummary = new CommonMetricsSnapshot();
 
 	public DescriptiveStatisticsHistogramStatistics(DescriptiveStatistics latencyHistogram) {
-		this.descriptiveStatistics = latencyHistogram;
+		latencyHistogram.apply(statisticsSummary);
 	}
 
 	@Override
 	public double getQuantile(double quantile) {
-		return descriptiveStatistics.getPercentile(quantile * 100);
+		return statisticsSummary.getPercentile(quantile * 100);
 	}
 
 	@Override
 	public long[] getValues() {
-		return Arrays.stream(descriptiveStatistics.getValues()).mapToLong(i -> (long) i).toArray();
+		return Arrays.stream(statisticsSummary.getValues()).mapToLong(i -> (long) i).toArray();
 	}
 
 	@Override
 	public int size() {
-		return (int) descriptiveStatistics.getN();
+		return (int) statisticsSummary.getCount();
 	}
 
 	@Override
 	public double getMean() {
-		return descriptiveStatistics.getMean();
+		return statisticsSummary.getMean();
 	}
 
 	@Override
 	public double getStdDev() {
-		return descriptiveStatistics.getStandardDeviation();
+		return statisticsSummary.getStandardDeviation();
 	}
 
 	@Override
 	public long getMax() {
-		return (long) descriptiveStatistics.getMax();
+		return (long) statisticsSummary.getMax();
 	}
 
 	@Override
 	public long getMin() {
-		return (long) descriptiveStatistics.getMin();
+		return (long) statisticsSummary.getMin();
+	}
+
+	/**
+	 * Function to extract several commonly used metrics in an optimised way, i.e. with as few runs
+	 * over the data / calculations as possible.
+	 *
+	 * <p>Note that calls to {@link #evaluate(double[])} or {@link #evaluate(double[], int, int)}
+	 * will not return a value but instead populate this class so that further values can be
+	 * retrieved from it.
+	 */
+	private static class CommonMetricsSnapshot implements UnivariateStatistic {
+		long count = 0;
+		double min = Double.NaN;
+		double max = Double.NaN;
+		double mean = Double.NaN;
+		double stddev = Double.NaN;
+		private Percentile percentilesImpl = new Percentile().withNaNStrategy(NaNStrategy.FIXED);
+
+		@Override
+		public double evaluate(final double[] values) throws MathIllegalArgumentException {
+			return evaluate(values, 0, values.length);
+		}
+
+		@Override
+		public double evaluate(double[] values, int begin, int length)
+				throws MathIllegalArgumentException {
+			this.count = length;
+			percentilesImpl.setData(values, begin, length);
+
+			SimpleStats secondMoment = new SimpleStats();
+			secondMoment.evaluate(values, begin, length);
+			this.mean = secondMoment.getMean();
+			this.min = secondMoment.getMin();
+			this.max = secondMoment.getMax();
+
+			this.stddev = new StandardDeviation(secondMoment).getResult();
+
+			return Double.NaN;
+		}
+
+		@Override
+		public CommonMetricsSnapshot copy() {
```

Comment:
is this method actually used somewhere?

Reply:
We need to implement this from the interface and honestly don't know whether this maybe used in, for example this:
```
	public DescriptiveStatisticsHistogramStatistics(DescriptiveStatistics latencyHistogram) {
		latencyHistogram.apply(statisticsSummary);
	}
```

But good point, I forgot to override this for `SimpleStats`.